<launch>
  <arg name="REMOTE_TF" default="/tf_low_frequency"/>
  <arg name="USE_INTERACTIVE_JOINT" default="true" />
  <arg name="USE_INTERACTIVE_ENDCOORDS" default="true" />
  <arg name="ROBOT_DESCRIPTION_REMOTE" default="robot_description" />
  <arg name="LAUNCH_DYNAMIC_TF_PUBLISHER" default="false" />
  <arg name="ROBOT" default="PR2" />
  <arg name="IS_PR2" default="true" />
  <arg name="ROBOT_ROOTFRAME" default="base_footprint" />

  <!-- joint marker -->
  <group if="$(arg USE_INTERACTIVE_JOINT)">
    <remap from="/tf" to="$(arg REMOTE_TF)"/>
    <remap from="/set_dynamic_tf" to="/dynamic_tf_publisher_remote/set_dynamic_tf"/>
    <remap from="/delete_tf" to="/dynamic_tf_publisher_remote/delete_tf"/>
    <remap from="/dissoc_tf" to="/dynamic_tf_publisher_remote/dissoc_tf"/>
    <remap from="/assoc_tf" to="/dynamic_tf_publisher_remote/assoc_tf"/>

    <!-- use joint_state_publisher instaed of dynamic_tf_publisher -->
    
    <!-- use joint_state_publisher when using pr2 -->
    <group if="$(arg IS_PR2)">
      <node pkg="robot_state_publisher" name="robot_state_publisher_for_urdf_marker" type="robot_state_publisher" >
	<remap from="joint_states" to="jsk_model_marker_interface/$(arg ROBOT)/joint_states_full_100hz" />
	<param name="~tf_prefix" type="string" value="$(arg ROBOT)" />
      </node>

      <node pkg="joint_state_publisher" name="joint_state_publisher_for_urdf_marker" type="joint_state_publisher" >
	<remap from="joint_states" to="jsk_model_marker_interface/$(arg ROBOT)/joint_states_full_100hz" />
	<rosparam>
          source_list: ['jsk_model_marker_interface/PR2/joint_states_100hz']
        </rosparam>
      </node>
    </group>

    <!-- not use joint_state_publisher otherwise -->
    <group unless="$(arg IS_PR2)">
      <node pkg="robot_state_publisher" name="robot_state_publisher_for_urdf_marker" type="robot_state_publisher" >
	<remap from="joint_states" to="jsk_model_marker_interface/$(arg ROBOT)/joint_states_100hz" />
	<param name="~tf_prefix" type="string" value="$(arg ROBOT)" />
      </node>
    </group>


    <node pkg="jsk_interactive_marker" type="joint_state_publisher.py" name="joint_state_publisher_for_jsk_interactive_marker" >
      <remap from="joint_states_sub" to="jsk_model_marker_interface/$(arg ROBOT)/joint_states" />
      <remap from="joint_states_pub" to="jsk_model_marker_interface/$(arg ROBOT)/joint_states_100hz" />
    </node>

    <node pkg="tf" type="static_transform_publisher" name="static_transform_publisher" args="0 0 0 0 0 0 $(arg ROBOT_ROOTFRAME) $(arg ROBOT)/$(arg ROBOT_ROOTFRAME) 100" />

    <node pkg="rviz" type="rviz" name="rviz_for_interactive_marker" args="-d $(find jsk_interactive_marker)/launch/atlas_joint_marker.rviz" respawn="true">
      <remap from="/robot_description" to="$(arg ROBOT_DESCRIPTION_REMOTE)"/>
    </node>

    <include file="$(find jsk_interactive_marker)/launch/urdf_model_markers.launch">
      <arg name="LAUNCH_DYNAMIC_TF_PUBLISHER" value="$(arg LAUNCH_DYNAMIC_TF_PUBLISHER)"/>
      <arg name="models" value="$(find jsk_interactive_marker)/launch/models/PR2-real.yaml"/>
    </include>
  </group>

  <!-- hand marker -->
  <group if="$(arg USE_INTERACTIVE_ENDCOORDS)" >
    <remap from="/tf" to="$(arg REMOTE_TF)"/>
    <remap from="/set_dynamic_tf" to="/dynamic_tf_publisher_remote/set_dynamic_tf"/>
    <remap from="/dissoc_tf" to="/dynamic_tf_publisher_remote/dissoc_tf"/>
    <remap from="/delete_tf" to="/dynamic_tf_publisher_remote/delete_tf"/>
    <remap from="/assoc_tf" to="/dynamic_tf_publisher_remote/assoc_tf"/>
    <!--
    <include file="$(find jsk_interactive_marker)/launch/atlas_interactive_marker_practical.launch" >

    </include>
      -->

    <include file="$(find jsk_interactive_marker)/launch/pr2_interactive_marker.launch" />


    <!-- eus-ik-controller -->
    <!--
    <node pkg="roseus" type="roseus" name="ik_controller" args="$(find hrpsys_gazebo_atlas)/euslisp/atlas-ik-controller.l" output="screen">
      <param name="robot" value="$(arg ROBOT)"/>
    </node>
    -->
    <node pkg="roseus" type="roseus" name="ik_controller" args="$(find jsk_interactive_marker)/euslisp/ik-controller.l" output="screen">
      <param name="robot" value="$(arg ROBOT)"/>
    </node>

    <!-- eus-ik-server -->
    <node pkg="roseus" type="roseus" name="pr2_ik_server"
	  args="$(find jsk_ik_server)/euslisp/pr2-ik-server.l"/>
  </group>
</launch>
